### PR TITLE
Upload local datasets using cli

### PR DESCRIPTION
**Pull request by tall-josh**
The Highlighter SDK code that enables the functions in this PR to work is [HERE](https://gitlab.com/silverpond/products/highlighter_all/-/merge_requests/523). These docs are only valid once that MR is merged.

### CODE REVIEW

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+```
+
+### From a Coco json file
+
+If you have a coco dataset json file you can use the `--file` option.
+
+If needed you can use `--image-dir` to append an directory to the `file_name`s
+im the coco image records.
+
+Upon completion you will also have a duplica coco json file with the `.images[].file_
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:82]:
```suggestion
Upon completion you will also have a duplicate coco json file with the `.images[].file_
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+## Images And Observations
+
+If you have an existing labelled dataset you can upload the images and 
+the labels as follows.
+
+**Coco Bounding Box or Polygon segmentaton are the only formats supported from the CLI at the moment**
+
+### Prerequisites
+  
+This tutorial expects you:
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:99]:
```suggestion
This tutorial expects you have done the following:
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+  1. [Highlighter Credentials](./highlighter-credentials.md)
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:101]:
```suggestion
  1. Created [Highlighter Credentials](./highlighter-credentials.md)
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  2. Create a Data Source in the Highlighter Web UI, note the ID from the URL, eg:
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:102]:
```suggestion
  2. Created a Data Source in the Highlighter Web UI, note the ID from the URL, eg:
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  `https://compuglobalhypermeganet.highlighter.ai/data_source/DATASOURCE_ID`
+  4. Create the Object Classes on the same name in Highlighter
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:104]:
```suggestion
  4. Created the Object Classes on the same name in Highlighter
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  3. Create an Assessment Process in Highlighter, note the ID from the URL, eq:
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:105]:
```suggestion
  3. Created an Assessment Process in Highlighter, note the ID from the URL, eq:
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  `https://compuglobalhypermeganet.highlighter.ai/projects/PROCESS_ID`
+  5. Add the Object Classes to the Assessment Process
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:107]:
```suggestion
  5. Added the Object Classes to the Assessment Process
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+  6. Your Highlighter User Id
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:108]:
```suggestion
  6. Found your Highlighter User Id
```

**Review by simon-dev-ml — CHANGES_REQUESTED** [review]

File: content/docs/how-to-guides/highlighter-credentials.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+#### Option 2: Highlighter Profiles
+
+If you're dealing with multiple Highlighter accounts you can use
+a profiles yaml file in at `~/.highlighter-profiles.yaml`
+
+```yaml
+# ~/.highlighter-profiles.yaml
+
+profile_name:
```
Comment (rskew) [inline content/docs/how-to-guides/highlighter-credentials.md:52]:
```suggestion
compuglobalhypermeganet:
```

File: content/docs/how-to-guides/highlighter-credentials.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  api_token: d000d0d0d0d0d0d0d000d0d0d0d0d0d0
+  endpoint_url: https://compuglobalhypermeganet.highlighter.ai/graphql
+
+...
+```
+
+You can now use the `--profile` option when using the CLI or
```
Comment (rskew) [inline content/docs/how-to-guides/highlighter-credentials.md:59]:
```suggestion
You can now use the `--profile compuglobalhypermeganet` option when using the CLI or
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+description = "How to upload images and optionally annotations to Highlighter using the CLI"
+date = 2024-03-12T08:00:00+00:00
+updated = 2024-03-12T08:00:00+00:00
+draft = false
+weight = 100
+sort_by = "weight"
+template = "docs/page.html"
+
+[extra]
+lead = "How to upload images and optionally annotations to Highlighter using the CLI"
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:12]: I don't think this line adds much, the title and first paragraph sum it up pretty well I think:

screenshot641

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+both options.
+
+
+## Images Only
+
+### Prerequisites
+  
+This tutorial expects you have done the following:
+
+  1. Have [Highlighter Credentials →](../highlighter-credentials/)
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:28]:
```suggestion
  1. Created [Highlighter Credentials →](../highlighter-credentials/)
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+  2. Created a Data Source in the Highlighter Web UI, note the ID from the URL, eg:
+  `https://compuglobalhypermeganet.highlighter.ai/data_source/DATASOURCE_ID`
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:30]:
```suggestion
  `https://your-account-subdomain.highlighter.ai/data_source/DATASOURCE_ID`
```
Non-blocking. Nothing against `compuglobalhypermeganet` but it takes the focus away from the `DATASOURCE_ID` when reading

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+
+## Images Only
+
+### Prerequisites
+  
+This tutorial expects you have done the following:
+
+  1. Have [Highlighter Credentials →](../highlighter-credentials/)
+  2. Created a Data Source in the Highlighter Web UI, note the ID from the URL, eg:
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:29]:
```suggestion
  2. Created a Data Source in the Highlighter Web UI, note the Data Source ID from the URL, eg:
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+  `https://compuglobalhypermeganet.highlighter.ai/data_source/DATASOURCE_ID`
+
+
+### Upload Images From a Directory
+
+The simplest form is just uploading a directory of files.
+
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:36]:
```suggestion
```
Non-blocking suggestion. I don't think this line is needed

File: content/docs/how-to-guides/upload-data-using-cli.md (hunk 11 above)
Comment (rskew) [reply content/docs/how-to-guides/upload-data-using-cli.md:12]:
```suggestion
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+The following will upload all files in `--image-dir` to the `--data-source-id`
+provided.
+
+Upon completion you will also have a local directory with symlinks mapping the
+newly created Highlighter file ids to the original files you uploaded.
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:41]: Where is this new local directory with symlinks?

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+```shell
+hl image create --data-source-id 123 --image-dir my_images/
+```
+
+
+### Upload Images From a Text File
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:48]:
```suggestion
### Upload Images Listed in a Text File
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+```
+# images.txt
+images/00.jpg
+images/01.jpg
+images/02.jpg
+```
+
+```shell
+
+# Upload images with paths in file
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:68]:
```suggestion
```shell
# Upload images with paths in file
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+hl image create --data-source-id 123 --file images.txt
+
+# Upload images with paths foo/PATH_FROM_TXT_FILE
+hl image create --data-source-id 123 --file images.txt --image-dir foo
+```
+
+### From a Coco json file
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:75]:
```suggestion
### Upload a Dataset from a Coco JSON File
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+If you have a coco dataset json file you can use the `--file` option.
+
+If needed you can use `--image-dir` to append an directory to the `file_name`s
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:79]:
```suggestion
If needed you can use `--image-dir` to append a directory to the `file_name`s
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+in the coco image records.
+
+Upon completion you will also have a duplicate coco json file with the `.images[].file_
+name` values updated to use the newly create Highlighter file ids
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:83]:
```suggestion
Upon completion you will also have a duplicate coco json file with the `.images[].file_name` values updated to use the newly create Highlighter file ids
```
line break creates a space

screenshot642

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+
+```shell
+# Upload images with paths image_dir/PATH
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:86]:
```suggestion
# Upload images located at foo/image_dir/FILENAME with paths image_dir/FILENAME listed in coco.json
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+hl image create --data-source-id 123 --file coco.json --image-dir foo
+```
+
+## Images And Observations
+
+If you have an existing labelled dataset you can upload the images and 
+the labels as follows.
+
+**Coco Bounding Box or Polygon segmentaton are the only formats supported from the CLI at the moment**
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:95]:
```suggestion
**Note: Coco Bounding Box or Polygon segmentaton are the only formats supported from the CLI at the moment**
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+
+### Prerequisites
+  
+This tutorial expects you have done the following:
+
+  1. Have [Highlighter Credentials](./highlighter-credentials.md)
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:101]:
```suggestion
  1. Created [Highlighter Credentials](./highlighter-credentials.md)
```

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+  2. Created a Data Source in the Highlighter Web UI, note the ID from the URL, eg:
+  `https://compuglobalhypermeganet.highlighter.ai/data_source/DATASOURCE_ID`
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:103]:
```suggestion
  `https://your-account-subdomain.highlighter.ai/data_source/DATASOURCE_ID`
```
as above

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+  4. Create the Object Classes of the same name in Highlighter
+  3. Created an Assessment Process in Highlighter, note the ID from the URL, eq:
+  `https://compuglobalhypermeganet.highlighter.ai/projects/PROCESS_ID`
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:106]:
```suggestion
  `https://your-account-subdomain.highlighter.ai/projects/PROCESS_ID`
```
as above

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 27]
```diff
@@ -0,0 +1,137 @@
...
+```
+annotations/
+  train.json
+  ...
+images/
+  ...
+```
+
+```shell
+# Upload the images, after this you should have a duplicate of train.json
```
Comment (rskew) [inline content/docs/how-to-guides/upload-data-using-cli.md:123]:
```suggestion
# Upload the images. After this you should have a duplicate of train.json
```

**Review by rskew — COMMENTED** [review]
Great work!

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+```
+
+```shell
+# Upload the images. After this you should have a duplicate of train.json
+# containing the new Highlighter file ids train_with_hl_ids.json
+hl image create --data-source-id 123 \
+  --file annotations/train.json \
+  --image-dir images
+
+# Upload the labels in the dataset to an Assessment Process
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:124]: Need to add docs before this line to create the object classes and add them to the process

Comment (simon-dev-ml) [reply content/docs/how-to-guides/upload-data-using-cli.md:124]: @tall-josh Let me know once you've done this and I can re-review

File: content/docs/how-to-guides/upload-data-using-cli.md [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+in the coco image records.
+
+Upon completion you will also have a duplicate coco json `<COCO_FILENAME>_with_hl_ids.json` file with the `.images[].file_name` values updated to use the newly created Highlighter file ids
+
+```shell
+# Upload images located at foo/image_dir/FILENAME with paths image_dir/FILENAME listed in coco.json
+hl image create --data-source-id 123 --file coco.json --image-dir foo
+```
+
+## Images And Observations
```
Comment (simon-dev-ml) [inline content/docs/how-to-guides/upload-data-using-cli.md:85]:
```suggestion
## Upload Images And Observations
```

**Review by simon-dev-ml — CHANGES_REQUESTED** [review]
Just few more little things and my comment [here](https://github.com/silverpond/highlighter-docs/pull/15#discussion_r1524162416), apart from that looking good

**Review by simon-dev-ml — APPROVED** [review]

**Pull request merged**